### ISSUE TITLE
LRScheduler doesn't use _get_closed_form_lr when initalizing scheduler

### ISSUE THREAD

**Issue by NST-d** [pytorch/pytorch#184517]
### 🐛 Describe the bug

When defining a class that inherits from `LRScheduler `, it can't be done by defining only  a `_get_closed_form_lr()` method, as in the initialization `_update_lr()` only uses `get_lr()`. I think it should also use the closed form.

A minimal example that uses a non multiplicative lambda schedule:

```
import torch
from typing import Callable

class BetterLambdaLRScheduler(torch.optim.lr_scheduler.LRScheduler):
    def __init__(self, optimizer: torch.optim.Optimizer, lr_lambda: Callable[[int], float | torch.Tensor], last_epoch: int =-1):
    self.lr_lambda = lr_lambda        
    super().__init__(optimizer, last_epoch=last_epoch)
        

    def _get_closed_form_lr(self) -> list[float | torch.Tensor]:
        return [
            self.lr_lambda(self.last_epoch)
            for _ in self.base_lrs
        ]
```

### Versions

Collecting environment information...
PyTorch version: 2.11.0+cu130
Is debug build: False
CUDA used to build PyTorch: 13.0
ROCM used to build PyTorch: N/A

OS: Ubuntu 22.04.3 LTS (x86_64)
GCC version: (Ubuntu 11.4.0-1ubuntu1~22.04.3) 11.4.0
Clang version: Could not collect
CMake version: version 3.22.1
Libc version: glibc-2.35

Python version: 3.12.11 (main, Jul 23 2025, 00:34:44) [Clang 20.1.4 ] (64-bit runtime)
Python platform: Linux-6.8.0-100-generic-x86_64-with-glibc2.35
Is CUDA available: True
CUDA runtime version: Could not collect
CUDA_MODULE_LOADING set to: 
GPU models and configuration: GPU 0: NVIDIA A100-SXM4-40GB
Nvidia driver version: 590.48.01
cuDNN version: Could not collect
Is XPU available: False
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True
Caching allocator config: N/A

CPU:
Architecture:                            x86_64
CPU op-mode(s):                          32-bit, 64-bit
Address sizes:                           43 bits physical, 48 bits virtual
Byte Order:                              Little Endian
CPU(s):                                  256
On-line CPU(s) list:                     0-255
Vendor ID:                               AuthenticAMD
Model name:                              AMD EPYC 7742 64-Core Processor
CPU family:                              23
Model:                                   49
Thread(s) per core:                      2
Core(s) per socket:                      64
Socket(s):                               2
Stepping:                                0
Frequency boost:                         enabled
CPU max MHz:                             2250.0000
CPU min MHz:                             1500.0000
BogoMIPS:                                4499.52
Flags:                                   fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush mmx fxsr sse sse2 ht syscall nx mmxext fxsr_opt pdpe1gb rdtscp lm constant_tsc rep_good nopl nonstop_tsc cpuid extd_apicid aperfmperf rapl pni pclmulqdq monitor ssse3 fma cx16 sse4_1 sse4_2 x2apic movbe popcnt aes xsave avx f16c rdrand lahf_lm cmp_legacy svm extapic cr8_legacy abm sse4a misalignsse 3dnowprefetch osvw ibs skinit wdt tce topoext perfctr_core perfctr_nb bpext perfctr_llc mwaitx cpb cat_l3 cdp_l3 hw_pstate ssbd mba ibrs ibpb stibp vmmcall fsgsbase bmi1 avx2 smep bmi2 cqm rdt_a rdseed adx smap clflushopt clwb sha_ni xsaveopt xsavec xgetbv1 xsaves cqm_llc cqm_occup_llc cqm_mbm_total cqm_mbm_local clzero irperf xsaveerptr rdpru wbnoinvd amd_ppin arat npt lbrv svm_lock nrip_save tsc_scale vmcb_clean flushbyasid decodeassists pausefilter pfthreshold avic v_vmsave_vmload vgif v_spec_ctrl umip rdpid overflow_recov succor smca sev sev_es ibpb_exit_to_user
Virtualization:                          AMD-V
L1d cache:                               4 MiB (128 instances)
L1i cache:                               4 MiB (128 instances)
L2 cache:                                64 MiB (128 instances)
L3 cache:                                512 MiB (32 instances)
NUMA node(s):                            2
NUMA node0 CPU(s):                       0-63,128-191
NUMA node1 CPU(s):                       64-127,192-255
Vulnerability Gather data sampling:      Not affected
Vulnerability Indirect target selection: Not affected
Vulnerability Itlb multihit:             Not affected
Vulnerability L1tf:                      Not affected
Vulnerability Mds:                       Not affected
Vulnerability Meltdown:                  Not affected
Vulnerability Mmio stale data:           Not affected
Vulnerability Reg file data sampling:    Not affected
Vulnerability Retbleed:                  Mitigation; untrained return thunk; SMT enabled with STIBP protection
Vulnerability Spec rstack overflow:      Mitigation; Safe RET
Vulnerability Spec store bypass:         Mitigation; Speculative Store Bypass disabled via prctl
Vulnerability Spectre v1:                Mitigation; usercopy/swapgs barriers and __user pointer sanitization
Vulnerability Spectre v2:                Mitigation; Retpolines; IBPB conditional; STIBP always-on; RSB filling; PBRSB-eIBRS Not affected; BHI Not affected
Vulnerability Srbds:                     Not affected
Vulnerability Tsa:                       Not affected
Vulnerability Tsx async abort:           Not affected
Vulnerability Vmscape:                   Mitigation; IBPB before exit to userspace

Versions of relevant libraries:
[pip3] Could not collect
[conda] Could not collect

**Comment by janeyx99** [comment]
Thanks for the issue. I'm curious what's the motivation behind this? In this case, would a workaround be to just have get_lr call _get_closed_form_lr()? In general, it is weird to override only a private method on a subclass.

**Comment by NST-d** [comment]
Yes, i think just using get_lr() to call _get_closed_form_lr() or only using get_lr should work fine, as far as i see it the base class never makes any assumptions that get_lr doesn't use a closed form. I just saw that some of the schedulers have implemented the _get_closed_form_lr, did so as well and stumbled into this bug. But it is curious why this option exists in the first place, as it basically changes the role of get_lr from updating the learning rate to just initalization, depending on the existance of the other method.

**Comment by janeyx99** [comment]
Ah I don't have a good reason, but: this was because we had at some point wanted to update our LRSchedulers to also allow closed form, but that migration never completed as our maintainer left. So we're left in a more hodge-podge state. @diputs-sudo has been looking into revamping our LRScheduler to be more stateless.

**Comment by diputs-sudo** [comment]
I think the surprising part is that `_get_closed_form_lr()` looks like an alternate scheduler definition API, but in practice the current contract still treats `get_lr()` as the canonical chainable/stateful path.

So even though some schedulers implement both, the closed form path ends up behaving more like optional support for `step(epoch=...)` rather than a fully independent scheduler definition

I could see a fallback to the closed form path when `get_lr()` isn’t overridden, although that probably gets tricky since some schedulers intentionally distinguish recursive updates from closed form recomputation.

So this feels less like a missing one line fallback and more like an artifact of the current scheduler abstraction being partly recursive/stateful and partly closed form.
